### PR TITLE
Clean and document ball filter

### PR DESCRIPTION
**Pull request by dstansby**
Here's some general cleaning and documenting of the ball filter part of cell detection. Please be critical and don't hesitate to add suggestions on review, would be good to get a close second pair of eyes on this.

### CODE REVIEW

File: src/cellfinder_core/tools/array_operations.py [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 
 # WARNING: skewed if does not fall exactly (integer multiple)
-def bin_array(arr, bin_width=15, bin_height=15):
+def bin_array(arr: np.ndarray, bin_width: int, bin_height: int):
```
Comment (alessandrofelder) [inline src/cellfinder_core/tools/array_operations.py:7]: Are we removing the default values for width and height here intentionally? (Would make sense given that bin_width _should_ typically be a divisor of the array shape so it's hard to have a sensible default, but wanted to double-check?)

File: src/cellfinder_core/tools/array_operations.py [hunk 2]
```diff
@@ -26,12 +26,17 @@ def bin_array(arr, bin_width=15, bin_height=15):
     return binned_array
 
 
-def pad_and_bin_mean(arr, bin_width=15, bin_height=15):
+def pad_and_bin_mean(arr: np.ndarray, bin_width: int, bin_height: int):
```
Comment (alessandrofelder) [inline src/cellfinder_core/tools/array_operations.py:29]: Intentional removal of defaults? (See comment above)

File: src/cellfinder_core/detect/filters/volume/ball_filter.py [hunk 3]
```diff
@@ -187,30 +215,60 @@ def _cube_overlaps(
...
+    Scan through *volume*, and mark pixels where there are enough surrounding
+    pixels with high enough intensity.
+
+    The surrounding area is defined by the *kernel*.
+
+    Parameters
+    ----------
+    max_height, max_width :
+        Maximum offsets for the ball filter.
+    good_tiles_mask :
```
Comment (alessandrofelder) [inline src/cellfinder_core/detect/filters/volume/ball_filter.py:258]: In line with #114 we should rename this to `brain_tiles_mask` or `bright_tiles_mask`.

File: src/cellfinder_core/tools/array_operations.py [hunk 4]
```diff
@@ -4,7 +4,7 @@
 
 
 # WARNING: skewed if does not fall exactly (integer multiple)
```
Comment (alessandrofelder) [inline src/cellfinder_core/tools/array_operations.py:6]: should we add a docstring to `bin_array()` and mention this there instead?

File: src/cellfinder_core/tools/array_operations.py (hunk 2 above)
Comment (alessandrofelder) [reply src/cellfinder_core/tools/array_operations.py:29]: Also needs docstring!

**Review by alessandrofelder — CHANGES_REQUESTED** [review]
I think some functions are still missing docstrings + a suggestion to rename "bin"?

File: src/cellfinder_core/detect/filters/volume/ball_filter.py [hunk 5]
```diff
@@ -29,26 +31,36 @@ def __init__(
...
-            y_upscale_factor * ball_xy_size,
-            z_upscale_factor * ball_z_size,
-        ]
-        tmp_ball_centre_position = [
-            np.floor(d / 2) for d in temp_kernel_shape
-        ]  # z_centre is xy_centre before resize
-        tmp_ball_radius = temp_kernel_shape[0] / 2.0
-        tmp_kernel = make_sphere(
-            temp_kernel_shape, tmp_ball_radius, tmp_ball_centre_position
+        # Create a spherical kernel.
```
[the remark this answers is not in the document]
Comment (dstansby) [reply src/cellfinder_core/detect/filters/volume/ball_filter.py:34]: 👍 added, comments welcome!

File: src/cellfinder_core/tools/array_operations.py (hunk 1 above)
Comment (dstansby) [reply src/cellfinder_core/tools/array_operations.py:7]: Yes, removing defaults was deliberate. 👍 I've added a docstring

File: src/cellfinder_core/detect/filters/volume/ball_filter.py (hunk 3 above)
Comment (dstansby) [reply src/cellfinder_core/detect/filters/volume/ball_filter.py:258]: 👍 done

File: src/cellfinder_core/tools/array_operations.py (hunk 4 above)
Comment (dstansby) [reply src/cellfinder_core/tools/array_operations.py:6]: I've added notes sections to `bin_array()`, `pad_and_bin_mean()`, and `bin_mean_3d()` to clarify this (and removed the warning comment)

File: src/cellfinder_core/tools/array_operations.py [hunk 6]
```diff
@@ -26,12 +26,17 @@ def bin_array(arr, bin_width=15, bin_height=15):
...
 
 
-def pad_and_bin_mean(arr, bin_width=15, bin_height=15):
+def pad_and_bin_mean(arr: np.ndarray, bin_width: int, bin_height: int):
     binned_array = bin_array(arr, bin_width, bin_height)
     return binned_array.mean(axis=(1, 3))
 
 
-def bin_mean_3d(arr, bin_width, bin_height, bin_depth):
+def bin_mean_3d(
```
[the remark this answers is not in the document]
Comment (dstansby) [reply src/cellfinder_core/tools/array_operations.py:34]: I think binning is the right word - returning the sum of values across a bin is exactly what a histogram does. See also [`scipy.stats.binned_statistic`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.binned_statistic.html), which is a 1D version of the code in `array_operations.py` here.

**Comment by dstansby** [comment]
All done, thanks for the suggestions! Worth double checking the new docstrings I've added.

File: src/cellfinder_core/detect/filters/volume/ball_filter.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (alessandrofelder) [reply src/cellfinder_core/detect/filters/volume/ball_filter.py:67]: Looks great!

**Review by alessandrofelder — APPROVED** [review]
Great docstrings 😍 - thank you.

As a minor consideration, I think the naming of `bin_array()`, `pad_and_bin_mean()` and `bin_mean_3d()` is inconsistent (the function name for 2d and 3d binned mean should differ only by a number?) and could be more expressive (I do agree "bin" is the right word now!) I'd suggest `get_2d_bins()`, `binned_mean_2d()` and `binned_mean_3d()` - the latter two names would match the `scipy` naming choice of `binned_statistic` referenced in the review reply - or similar names instead?

I am happy for this to be merged without my nitpicking over function names though!

**Pull request merged**